### PR TITLE
Bug 1951084: remove mounts.conf again, but patch buildah change needed to make transient mounts work for us

### PR DESCRIPTION
**Pull request by gabemontero**
@nalind @bparees the "meets min" change to buildah that allows host entitlements to work for openshift builds, but without a mounts.conf file in the builder image, and warning message when we run on rhcos

https://github.com/containers/buildah/pull/3117 is a slightly different take

I'll next see about porting that in, and verify on our UPI with entitlements cluster

### CODE REVIEW

**Comment by gabemontero** [comment]
> @nalind @bparees the "meets min" change to buildah that allows host entitlements to work for openshift builds, but without a mounts.conf file in the builder image, and warning message when we run on rhcos
> 
> [containers/buildah#3117](https://github.com/containers/buildah/pull/3117) is a slightly different take
> 
> I'll next see about porting that in, and verify on our UPI with entitlements cluster

I was able to pull in the https://github.com/containers/buildah/pull/3117 version of the buildah change and also verify that host entitlements still work

**Comment by gabemontero** [comment]
I've redone this PR to vendor in the specific buildah commit that pulls in the change I verified in https://github.com/openshift/builder/pull/239#issuecomment-811309123

**Review by adambkaplan — COMMENTED** [review]
/hold

We are rebasing buildah in a separate PR, waiting for this to merge first.

**Comment by gabemontero** [comment]
> /hold
> 
> We are rebasing buildah in a separate PR, waiting for this to merge first.

Yep was just thinking this AM that when the 1.20.0 bump merges, this PR could lose its bump buildah commit

**Comment by gabemontero** [comment]
> > /hold
> > We are rebasing buildah in a separate PR, waiting for this to merge first.
> 
> Yep was just thinking this AM that when the 1.20.0 bump merges, this PR could lose its bump buildah commit

I was incorrect.  https://github.com/containers/buildah/pull/3117 is *NOT* in v1.20.0

That said, still need to rebase, then add new bump.

File: vendor/github.com/containers/buildah/run_linux.go [hunk 1]
```diff
@@ -519,8 +519,9 @@ func (b *Builder) setupMounts(mountPoint string, spec *specs.Spec, bundlePath st
 		return err
 	}
 
+	allMounts := util.SortMounts(append(append(append(append(append(volumes, builtins...), secretMounts...), bindFileMounts...), specMounts...), sysfsMount...))
```
Comment (gabemontero) [inline vendor/github.com/containers/buildah/run_linux.go:522]: this is the buildah fix we need that is not in the latest buildah tag, v1.20.0

**Comment by gabemontero** [comment]
/assign @adambkaplan 

minimally for approval

please let me know @adambkaplan if you just want to approve this PR and have me solicit review from other members of @openshift/openshift-team-build-api 

thanks

**Comment by gabemontero** [comment]
@xiuwang - verification for this will entail
1) running a build on normal RHCOS workers to verify the WARNING messages in the BZ do not appear
2) running a build using entitlements on RHEL worker nodes like what @wewang58 did for https://bugzilla.redhat.com/show_bug.cgi?id=1940488 and make sure those still work

File: pkg/build/builder/daemonless.go [hunk 2]
```diff
@@ -309,10 +317,85 @@ func buildDaemonlessImage(sc types.SystemContext, store storage.Store, isolation
...
+		return mounts
+	}
+	fs := s2ifs.NewFileSystem()
+	err = fs.CopyContents(path, tmpDir, map[string]string{})
+	if err != nil {
+		log.V(0).Infof("Red Hat subscription content will not be available in this build: failed to copy %s secrets: %v", pathEnd, err)
+		return mounts
+	}
+	mounts = append(mounts, fmt.Sprintf("%s:/run/secrets/%s:rw,nodev,noexec,nosuid", tmpDir, pathEnd))
+	return mounts
```
Comment (nalind) [inline pkg/build/builder/daemonless.go:386]:
```suggestion
```

Comment (gabemontero) [reply pkg/build/builder/daemonless.go:386]: hmm ... curious that even compiled ... thanks @nalind

File: pkg/build/builder/daemonless.go [hunk 3]
```diff
@@ -309,10 +317,84 @@ func buildDaemonlessImage(sc types.SystemContext, store storage.Store, isolation
...
+	mounts = append(mounts, fmt.Sprintf("%s:/run/secrets/%s:rw,nodev,noexec,nosuid", filepath.Join(tmpDir, repoFile), repoFile))
+	return mounts
+}
+
+func coreAppendSecretLinksToDirs(pathStart, pathEnd string, mounts []string) []string {
+	path := filepath.Join(pathStart, pathEnd)
+	st, err := os.Stat(path)
+	if err != nil {
+		// since the presence of dir secret is not a given, we won't log this a V(0)
+		log.V(0).Infof("Red Hat subscription content will not be available in this build: failed to stat directory %s: %v", path, err)
```
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:363]: Meant to log this at V(4)?

File: pkg/build/builder/daemonless.go [hunk 4]
```diff
@@ -309,10 +317,84 @@ func buildDaemonlessImage(sc types.SystemContext, store storage.Store, isolation
...
+		return mounts
+	}
+	if !st.IsDir() && st.Mode()&os.ModeSymlink == 0 {
```
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go:366]: space for readability? (looks like we are doing a bitwise AND here)

```suggestion
	if !st.IsDir() && (st.Mode() & os.ModeSymlink == 0) {
```

go fmt may remove my parens in the suggestion

File: go.mod [hunk 5]
```diff
@@ -5,8 +5,8 @@ go 1.13
 require (
 	github.com/MakeNowJust/heredoc v0.0.0-20170808103936-bb23615498cd
 	github.com/Nvveen/Gotty v0.0.0-20120604004816-cd527374f1e5 // indirect
-	github.com/containers/buildah v1.20.0
-	github.com/containers/common v0.35.3
+	github.com/containers/buildah v1.20.1-0.20210331202609-d29b04dba7fa
```
Comment (adambkaplan) [inline go.mod:8]: As @nalind pointed out in slack, we now have a v1.20.1 tag upstream.

**Review by adambkaplan — COMMENTED** [review]
/approve

Mostly nits, otherwise in good shape.

File: pkg/build/builder/daemonless.go (hunk 2 above)
Comment (gabemontero) [reply pkg/build/builder/daemonless.go:386]: fix pushed

File: pkg/build/builder/daemonless.go (hunk 3 above)
Comment (gabemontero) [reply pkg/build/builder/daemonless.go:363]: part of next push

File: pkg/build/builder/daemonless.go (hunk 4 above)
Comment (gabemontero) [reply pkg/build/builder/daemonless.go:366]: go fmt leaves the parens but removes the spaces you added @adambkaplan 

`if !st.IsDir() && (st.Mode()&os.ModeSymlink == 0) {`

that ^^ will be part of next push

File: go.mod (hunk 5 above)
Comment (gabemontero) [reply go.mod:8]: part of next push

**Comment by gabemontero** [comment]
updates pushed @adambkaplan @nalind thanks 

note what `gofmt` in fact did do for  the one suggestion @adambkaplan

**Comment by gabemontero** [comment]
e2e-aws hit that new kubelet panic the node team is tracking (per recent slack messages I've seen)

/retest

**Comment by wewang58** [comment]
Launched a cluster and tested entitlement build, no warning anymore.
/lgtm

**Comment by wewang58** [comment]
e2e-aws job failed with clusteroperator/ingress should not change condition/Available

**Pull request merged**